sdk/core/azure-core/inc/http/bodyStream.hpp

```diff
@@ -0,0 +1,269 @@
+    {
+    }
+
+    int64_t Length() const override { return this->m_length; }
+
+    int64_t Read(Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
+      // Copy what's left or just the count
+      std::memcpy(buffer, this->m_data + m_offset, (size_t)copy_length);
+      // move position
+      m_offset += copy_length;
+
+      return copy_length;
+    }
+
+    void Rewind() override { m_offset = 0; }
+
+    void Close() override {}
+  };
+
+  // Use for request with no body
+  class NullBodyStream : public Azure::Core::Http::BodyStream {
+  public:
+    explicit NullBodyStream() {}
+
+    ~NullBodyStream() override {}
+
+    int64_t Length() const override { return 0; }
+
+    void Rewind() override {}
+
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      (void)context;
+      (void)buffer;
+      (void)count;
+      return 0;
+    };
+
+    void Close() override {}
+  };
+
+#ifdef Posix
+  class FileBodyStream : public BodyStream {
+  private:
+    // in mutable
+    int m_fd;
+    int64_t m_baseOffset;
+    int64_t m_length;
+    // mutable
+    int64_t m_offset;
+
+  public:
+    FileBodyStream(int fd, int64_t offset, int64_t length)
+        : m_fd(fd), m_baseOffset(offset), m_length(length), m_offset(0)
+    {
+    }
+
+    ~FileBodyStream() override {}
+
+    // Rewind seek back to 0
+    void Rewind() override { this->m_offset = 0; }
+
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      auto result = pread(
+          this->m_fd,
+          buffer,
+          std::min(count, this->m_length - this->m_offset),
+          this->m_baseOffset + this->m_offset);
+      this->m_offset += result;
+      return result;
+    }
+
+    int64_t Length() const override { return this->m_length; };
+
+    // close does nothing opp
+    void Close() {}
+  };
+#endif
+
+#ifdef Windows
+  class FileBodyStream : public BodyStream {
+  private:
+    // in mutable
+    HANDLE m_hFile;
+    int64_t m_baseOffset;
+    int64_t m_length;
+    // mutable
+    int64_t m_offset;
+
+  public:
+    FileBodyStream(HANDLE hFile, int64_t offset, int64_t length)
+        : m_hFile(hFile), m_baseOffset(offset), m_length(length), m_offset(0)
+    {
+    }
+
+    ~FileBodyStream() override {}
+
+    // Rewind seek back to 0
+    void Rewind() override { this->m_offset = 0; }
+
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      DWORD numberOfBytesRead;
+      auto o = OVERLAPPED();
+      o.Offset = (DWORD)(this->m_baseOffset + this->m_offset);
+      o.OffsetHigh = (DWORD)((this->m_baseOffset + this->m_offset) >> 32);
+
+      auto result = ReadFile(
```

Comment:
error handling

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/237